### DISCUSSION TITLE
Maintain different Angular Versions in one pnpm Workspace

### DISCUSSION THREAD

**Discussion by GregOnNet** [pnpm/pnpm#8193] [Q&A]
Hey together,

currently, I am working on a demo, setting up a Micro Frontend Environment with Angular 16 and Angular 17, 18 apps.
This repository is about to be used for workshops and webinars.

My idea is to have everything in one repository to simplify code navigation for the workshop attendees.
Since every package uses different versions of the same package, I configured pnpm to let each package have its own lock-file.

```
# .npmrc
shared-workspace-lockfile = false
``` 

## The problem

After installing all dependencies with `pnpm install`, I start the frontends with `pnpm -r start`.
Unfortunatelly, packages cannot be resolved as expected.

```
✘ [ERROR] Cannot find package '~\ng-microfrontend\packages\remote-ng-16\node_modules\.pnpm\@angular+com…
│ Did you mean to import typescript@5.1.3/node_modules/typescript/lib/typescript.js? [plugin angular-compiler]
```

## Reproduction

- Clone [ng-microfrontends](https://github.com/GregOnNet/ng-microfrontend/tree/e4493efe923c859293628afd082731867fe9df98)
- Run `pnpm install`
- Run `pnpm -r start`
- Note that the apps do not start due to an error

![image](https://github.com/pnpm/pnpm/assets/444278/ebc07d20-e059-40ca-bc15-cdb33a4dc2fd)

## Question

- Do I try to misuse pnpm workspaces because maintaining multiple versions of the same package is not supported?
- Is there a way to set up the workspace as described above?

Thanks in advance for any hint. 🙏🏻

Cheers
Gregor

**Comment by zkochan** [comment]
I would recommend against using `shared-workspace-lockfile = false`. If angular is supposed to be a peer dependency, then you may use [injected dependencies](https://pnpm.io/package_json#dependenciesmetainjected), it will allow to use multiple angular versions in the same workspace.

You could also try using bit components, which have angular component ready to use: https://bit.dev/docs/angular-introduction/

Bit also has multiple angular versions support baked in and it uses pnpm under the hood to manager dependencies.
